### PR TITLE
new source: kafka

### LINKED ISSUE #2655: New Source: Kafka
The scope of the features should be supported for kafka source connector are:

1. Mapping onto airbyte output schema: kafka topic should be stream and fields might be free-form json value or parsed first level json objects. (additional properties = true).

- Data structure: one stream for each topic. For each stream we should have the following columns: partition, offset, timestamp and key with composite primary key partition and offset. The timestamp column may vary with create_time or log_append_time depending on the kafka config.
- Reading messages not one at a time but one batch (20 or 100 etc..) messages at a time. It’s much faster to do a batch than read one message at a time. 
- We need the ability to make the user decide whether we will do basic json extraction of first level Json elements. The unboxed messages must be in JSON format. For each first level Json field a new column needs to be created (with the correct field type like array, string, etc….)
- Column names must correspond to json first elements names.
- Unboxing means that we assume the schema of the messages is consistent across messages. Therefore we can do better than just saying the output is an object with additionalProperties = true. We can read the first N messages from the topic to discover the schema then output that from the discover() call. 

2. Authentication should be implemented within a TLS and SASL.
3. Airbyte should read and persist records in batch mode, within a fixed consumer group size (to avoid kafka partitions rebalancing).
4. Full refresh is supported with initial offset setting for a topic per partition.
5. Incremental refresh should be supported.
6. CDC option should be supported.

Phasing of implementation.
  
MVP scope:

1. Full refresh.
2. Possibility to configure topic offset per partition.
3. SASL auth    
4. Within batch processing with one consumer

Phase 2 scope:

1. tls auth
2. Incremental sync

Phase 3 scope:

1. Boxed unboxed json checkbox.	
2. CDC

Implementation Scoping

1. To connect to the kafka spring boot kafka template can be used.
2. Testing can be done via https://www.testcontainers.org/modules/kafka/
3. Integration test.
4. Acceptance test.
5. Comprehensive data test should be written. I suppose nulls, empty values, mail formatted json can be validated
6. Write a follow up ticket for fixing type conversion issues within a comprehensive test (ticket number)?
7. Implementation guideline:
8. Discover needed abstraction changes.
9. Setup test container
10. Write integration and acceptance test, Implement abstractions within a connector source code.
11. Write a comprehensive test.

### PR DESCRIPTION
**Pull request by heade**
## What
Source Kafka: https://github.com/airbytehq/airbyte/issues/2655

## How
 Implemented the connector logic for Kafka as a source for Airbyte.

MVP scope:
-Full refresh.
-Possibility to configure topic offset per partition.
-SASL auth
-Within batch processing with one consumer

Phase 2 scope:
-tls auth
-Incremental sync

2 types of subscriptions are implemented: topic-pattern and topic-partitions. To build streams in the first case, all available topics for this consumer are taken, and then filtered using a pattern: [link](https://github.com/airbytehq/airbyte/pull/5906/files#diff-3e0c7e09b3b80bf6201e5ed285d7695db17ba0ec30e9574ce41d9aa4e8f9f7c4R129)

## Recommended reading order
1. `spec.json `
2. `KafkaSourceConfig.java `
3. `KafkaSource.java`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [x] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [x] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by sherifnada** [comment]
@heade could you please include some more context for liren to be able to review the code? For example, links to any design docs, the reading order, any interesting implementation details? Also could you maximize the appropriate checklist section i.e: "New connector"?

File: airbyte-integrations/connectors/source-kafka/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+### Iteration
+TODO: _which commands should a developer use to run this connector locally?_
+
+### Testing
+#### Unit Tests
+TODO: _how can a user run unit tests?_
+
+#### Integration Tests
+TODO: _how can a user run integration tests?_
+_this section is currently under construction -- please reach out to us on Slack for help with setting up Airbyte's standard test suite_
```
Comment (tuliren) [inline airbyte-integrations/connectors/source-kafka/README.md:22]: This `TODO`s should either be resolved or removed.

**Review by tuliren — APPROVED** [review]
Nice! Thanks for the contribution.

A few docs and links are missing. But I can do that in a follow up PR.

**Review by tuliren — COMMENTED** [review]
@heade, sorry that I did not know you are from the GL team. Can you create a follow-up PR to complete the remaining tasks and publish this connector?

**Pull request merged**